File: lint-rules-rxjava2-lint/src/test/java/com/vanniktech/lintrules/rxjava2/RxJava2MissingCompositeDisposableClearDetectorTest.java
package com.vanniktech.lintrules.rxjava2;

import com.android.tools.lint.detector.api.Detector;
import com.android.tools.lint.detector.api.Issue;
import java.util.List;
import org.intellij.lang.annotations.Language;

import static com.vanniktech.lintrules.rxjava2.RxJava2MissingCompositeDisposableClearDetector.ISSUE_MISSING_COMPOSITE_DISPOSABLE_CLEAR;
import static java.util.Collections.singletonList;
import static org.fest.assertions.api.Assertions.assertThat;

public class RxJava2MissingCompositeDisposableClearDetectorTest extends RxJavaLintDetectorTest {
  public void testNoCompositeDisposable() throws Exception {
    @Language("JAVA") final String source = ""
        + "package foo;\n"
        + "import io.reactivex.disposables.CompositeDisposable;\n"
        + "public class Example {\n"
        + "}";
    assertThat(lintProject(stubCompositeDisposable, java(source))).isEqualTo(NO_WARNINGS);
  }

  public void testCompositeDisposableMissingClearSuppressed() throws Exception {
    @Language("JAVA") final String source = ""
        + "package foo;\n"
        + "import io.reactivex.disposables.CompositeDisposable;\n"
        + "import android.annotation.SuppressLint;\n"
        + "public class Example {\n"
        + "  @SuppressLint(\"MissingCompositeDisposableClear\") CompositeDisposable cd;\n"
        + "}";
    assertThat(lintProject(stubSuppressLint, stubCompositeDisposable, java(source))).isEqualTo(NO_WARNINGS);
  }

  public void testCompositeDisposableMissingClear() throws Exception {
    @Language("JAVA") final String source = ""
        + "package foo;\n"
        + "import io.reactivex.disposables.CompositeDisposable;\n"
        + "public class Example {\n"
        + "  CompositeDisposable cd;\n"
        + "}";
    assertThat(lintProject(stubCompositeDisposable, java(source))).isEqualTo("src/foo/Example.java:4: Error: clear() is not called. [MissingCompositeDisposableClear]\n"
        + "  CompositeDisposable cd;\n"
        + "  ~~~~~~~~~~~~~~~~~~~~~~~\n"
        + "1 errors, 0 warnings\n");
  }

  public void testDisposableMissingClearIsIgnored() throws Exception {
    @Language("JAVA") final String source = ""
        + "package foo;\n"
        + "import io.reactivex.disposables.Disposable;\n"
        + "public class Example {\n"
        + "  Disposable disposable;\n"
        + "}";
    assertThat(lintProject(stubDisposable, java(source))).isEqualTo(NO_WARNINGS);
  }

  public void testMultipleCompositeDisposableMissingClear() throws Exception {
    @Language("JAVA") final String source = ""
        + "package foo;\n"
        + "import io.reactivex.disposables.CompositeDisposable;\n"
        + "public class Example {\n"
        + "  CompositeDisposable cd1;\n"
        + "  CompositeDisposable cd2;\n"
        + "  CompositeDisposable cd3;\n"
        + "}";
    assertThat(lintProject(stubCompositeDisposable, java(source))).isEqualTo("src/foo/Example.java:4: Error: clear() is not called. [MissingCompositeDisposableClear]\n"
        + "  CompositeDisposable cd1;\n"
        + "  ~~~~~~~~~~~~~~~~~~~~~~~~\n"
        + "src/foo/Example.java:5: Error: clear() is not called. [MissingCompositeDisposableClear]\n"
        + "  CompositeDisposable cd2;\n"
        + "  ~~~~~~~~~~~~~~~~~~~~~~~~\n"
        + "src/foo/Example.java:6: Error: clear() is not called. [MissingCompositeDisposableClear]\n"
        + "  CompositeDisposable cd3;\n"
        + "  ~~~~~~~~~~~~~~~~~~~~~~~~\n"
        + "3 errors, 0 warnings\n"
        + "");
  }

  public void testCompositeDisposableHavingClear() throws Exception {
    @Language("JAVA") final String source = ""
        + "package foo;\n"
        + "import io.reactivex.disposables.CompositeDisposable;\n"
        + "public class Example {\n"
        + "  CompositeDisposable cd;\n"
        + "  public void foo() {\n"
        + "   cd.clear();\n"
        + "  }\n"
        + "}";
    assertThat(lintProject(stubCompositeDisposable, java(source))).isEqualTo(NO_WARNINGS);
  }

  public void testMultipleCompositeDisposableClear() throws Exception {
    @Language("JAVA") final String source = ""
        + "package foo;\n"
        + "import io.reactivex.disposables.CompositeDisposable;\n"
        + "public class Example {\n"
        + "  CompositeDisposable cd1;\n"
        + "  CompositeDisposable cd2;\n"
        + "  CompositeDisposable cd3;\n"
        + "  public void foo() {\n"
        + "   cd1.clear();\n"
        + "  }\n"
        + "  public void bar() {\n"
        + "   cd2.clear();\n"
        + "  }\n"
        + "}";
    assertThat(lintProject(stubCompositeDisposable, java(source))).isEqualTo("src/foo/Example.java:6: Error: clear() is not called. [MissingCompositeDisposableClear]\n"
        + "  CompositeDisposable cd3;\n"
        + "  ~~~~~~~~~~~~~~~~~~~~~~~~\n"
        + "1 errors, 0 warnings\n");
  }

  @Override protected Detector getDetector() {
    return new RxJava2MissingCompositeDisposableClearDetector();
  }

  @Override protected List<Issue> getIssues() {
    return singletonList(ISSUE_MISSING_COMPOSITE_DISPOSABLE_CLEAR);
  }
}

File: lint-rules-rxjava2-lint/src/test/java/com/vanniktech/lintrules/rxjava2/RxJava2MissingCompositeDisposableClearDetectorTest.kt
package com.vanniktech.lintrules.rxjava2

import com.android.tools.lint.checks.infrastructure.TestFiles.java
import com.android.tools.lint.checks.infrastructure.TestFiles.kt
import com.android.tools.lint.checks.infrastructure.TestLintTask.lint
import org.junit.Test

class RxJava2MissingCompositeDisposableClearDetectorTest {
  @Test fun noCompositeDisposable() {
    lint()
        .files(rxJava2(), java("""
          package foo;

          class Example {
          }""").indented())
        .issues(ISSUE_MISSING_COMPOSITE_DISPOSABLE_CLEAR)
        .run()
        .expectClean()
  }

  @Test fun compositeDisposableMissingClear() {
    lint()
        .files(rxJava2(), java("""
          package foo;

          import io.reactivex.disposables.CompositeDisposable;

          class Example {
            CompositeDisposable cd;
          }""").indented())
        .issues(ISSUE_MISSING_COMPOSITE_DISPOSABLE_CLEAR)
        .run()
        .expect("""
          |src/foo/Example.java:6: Error: clear() is not called. [RxJava2MissingCompositeDisposableClear]
          |  CompositeDisposable cd;
          |  ~~~~~~~~~~~~~~~~~~~~~~~
          |1 errors, 0 warnings""".trimMargin())
  }

  @Test fun disposableMissingDisposeIsIgnored() {
    lint()
        .files(rxJava2(), java("""
          package foo;

          import io.reactivex.disposables.Disposable;

          class Example {
            Disposable disposable;
          }""").indented())
        .issues(ISSUE_MISSING_COMPOSITE_DISPOSABLE_CLEAR)
        .run()
        .expectClean()
  }

  @Test fun multipleCompositeDisposableMissingClear() {
    lint()
        .files(rxJava2(), java("""
          package foo;

          import io.reactivex.disposables.CompositeDisposable;

          class Example {
            CompositeDisposable cd1;
            CompositeDisposable cd2;
          }""").indented())
        .issues(ISSUE_MISSING_COMPOSITE_DISPOSABLE_CLEAR)
        .run()
        .expect("""
          |src/foo/Example.java:6: Error: clear() is not called. [RxJava2MissingCompositeDisposableClear]
          |  CompositeDisposable cd1;
          |  ~~~~~~~~~~~~~~~~~~~~~~~~
          |src/foo/Example.java:7: Error: clear() is not called. [RxJava2MissingCompositeDisposableClear]
          |  CompositeDisposable cd2;
          |  ~~~~~~~~~~~~~~~~~~~~~~~~
          |2 errors, 0 warnings
          |""".trimMargin())
  }

  @Test fun multipleCompositeDisposableMissingClearInKotlin() {
    lint()
        .files(rxJava2(), kt("""
          package foo

          import io.reactivex.disposables.CompositeDisposable

          class Example {
            val cd1: CompositeDisposable
            val cd2: CompositeDisposable
          }""").indented())
        .issues(ISSUE_MISSING_COMPOSITE_DISPOSABLE_CLEAR)
        .run()
        .expect("""
          |src/foo/Example.kt:6: Error: clear() is not called. [RxJava2MissingCompositeDisposableClear]
          |  val cd1: CompositeDisposable
          |  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          |src/foo/Example.kt:7: Error: clear() is not called. [RxJava2MissingCompositeDisposableClear]
          |  val cd2: CompositeDisposable
          |  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          |2 errors, 0 warnings
          |""".trimMargin())
  }

  @Test fun compositeDisposableHavingClear() {
    lint()
        .files(rxJava2(), java("""
          package foo;

          import io.reactivex.disposables.CompositeDisposable;

          class Example {
            CompositeDisposable cd;
            public void foo() {
             cd.clear();
            }
          }""").indented())
        .issues(ISSUE_MISSING_COMPOSITE_DISPOSABLE_CLEAR)
        .run()
        .expectClean()
  }

  @Test fun compositeDisposableHavingClearInKotlin() {
    lint()
        .files(rxJava2(), kt("""
          package foo

          import io.reactivex.disposables.CompositeDisposable

          class Example {
            val cd = CompositeDisposable()
            fun foo() {
             cd.clear()
            }
          }""").indented())
        .issues(ISSUE_MISSING_COMPOSITE_DISPOSABLE_CLEAR)
        .run()
        .expectClean()
  }

  @Test fun missingOneCompositeDisposableClear() {
    lint()
        .files(rxJava2(), java("""
          package foo;

          import io.reactivex.disposables.CompositeDisposable;

          class Example {
            CompositeDisposable cd1;
            CompositeDisposable cd2;
            public void foo() {
             cd1.clear();
            }
          }""").indented())
        .issues(ISSUE_MISSING_COMPOSITE_DISPOSABLE_CLEAR)
        .run()
        .expect("""
          |src/foo/Example.java:7: Error: clear() is not called. [RxJava2MissingCompositeDisposableClear]
          |  CompositeDisposable cd2;
          |  ~~~~~~~~~~~~~~~~~~~~~~~~
          |1 errors, 0 warnings""".trimMargin())
  }

  @Test fun missingOneCompositeDisposableClearInKotlin() {
    lint()
        .files(rxJava2(), kt("""
          package foo

          import io.reactivex.disposables.CompositeDisposable

          class Example {
            val cd1 = CompositeDisposable()
            val cd2 = CompositeDisposable()
            fun foo() {
             cd1.clear()
            }
          }""").indented())
        .issues(ISSUE_MISSING_COMPOSITE_DISPOSABLE_CLEAR)
        .run()
        .expect("""
          |src/foo/Example.kt:7: Error: clear() is not called. [RxJava2MissingCompositeDisposableClear]
          |  val cd2 = CompositeDisposable()
          |  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          |1 errors, 0 warnings""".trimMargin())
  }

  @Test fun clearInIfStatement() {
    lint()
        .files(rxJava2(), java("""
        package foo;

        import io.reactivex.disposables.CompositeDisposable;

        class Example {
          CompositeDisposable cd;
          CompositeDisposable cd2;

          public void foo() {
            if (true) {
              cd.clear();
            }
          }

          public void foo2(){
            if (false) {

            } else {
              cd2.clear();
            }
          }
        }""").indented())
        .issues(ISSUE_MISSING_COMPOSITE_DISPOSABLE_CLEAR)
        .run()
        .expectClean()
  }

  @Test fun clearInIfStatementInKotlin() {
    lint()
        .files(rxJava2(), kt("""
        package foo

        import io.reactivex.disposables.CompositeDisposable

        class Example {
          val cd: CompositeDisposable
          val cd2: CompositeDisposable

          fun foo() {
            if (true) {
              cd.clear()
            }
          }

          fun foo2(){
            if (false) {

            } else {
              cd2.clear()
            }
          }
        }""").indented())
        .issues(ISSUE_MISSING_COMPOSITE_DISPOSABLE_CLEAR)
        .run()
        .expectClean()
  }
}
